rpc/types/events.go

```diff
@@ -0,0 +1,223 @@
+		var log evmtypes.Log
+		if err := json.Unmarshal(raw, &log); err != nil {
+			return nil, err
+		}
+
+		logs = append(logs, &log)
+	}
+	return evmtypes.LogsToEthereum(logs), nil
+}
+
+// ParsedTxs is the tx infos parsed from eth tx events.
+type ParsedTxs struct {
+	// one item per message
+	Txs []ParsedTx
+	// map tx hash to msg index
+	TxHashes map[common.Hash]int
+}
+
+// ParseTxResult parse eth tx infos from cosmos-sdk events.
+// It supports two event formats, the formats are described in the comments of the format constants.
+func ParseTxResult(result *abci.ResponseDeliverTx) (*ParsedTxs, error) {
+	format := eventFormatUnknown
+	// the index of current ethereum_tx event in format 1 or the second part of format 2
+	eventIndex := -1
+	var txs []ParsedTx
+	txHashes := make(map[common.Hash]int)
+	for _, event := range result.Events {
+		switch event.Type {
+		case evmtypes.EventTypeEthereumTx:
+			if format == eventFormatUnknown {
+				// discover the format version by inspect the first ethereum_tx event.
+				if len(event.Attributes) > 2 {
+					format = eventFormat1
+				} else {
+					format = eventFormat2
+				}
+			}
+
+			if len(event.Attributes) == 2 {
+				// the first part of format 2
+				msgIndex := len(txs)
+				tx := NewParsedTx(msgIndex)
+				if err := fillTxAttributes(&tx, event.Attributes); err != nil {
+					return nil, err
+				}
+				txs = append(txs, tx)
+				txHashes[tx.Hash] = msgIndex
+			} else {
+				// format 1 or second part of format 2
+				eventIndex++
+				if format == eventFormat1 {
+					// append tx
+					msgIndex := len(txs)
+					tx := NewParsedTx(msgIndex)
+					if err := fillTxAttributes(&tx, event.Attributes); err != nil {
+						return nil, err
+					}
+					txs = append(txs, tx)
+					txHashes[tx.Hash] = msgIndex
+				} else {
+					// the second part of format 2, update tx fields
+					if err := fillTxAttributes(&txs[eventIndex], event.Attributes); err != nil {
+						return nil, err
+					}
+				}
+			}
+		case evmtypes.EventTypeTxLog:
+			// reuse the eventIndex set by previous ethereum_tx event
+			txs[eventIndex].RawLogs = parseRawLogs(event.Attributes)
+		}
+	}
+
+	// some old versions miss some events, fill it with tx result
+	if len(txs) == 1 {
+		txs[0].GasUsed = uint64(result.GasUsed)
+	}
+
+	return &ParsedTxs{
+		Txs: txs, TxHashes: txHashes,
+	}, nil
+}
+
+// GetTxByHash find ParsedTx by tx hash, returns nil if not exists.
+func (p *ParsedTxs) GetTxByHash(hash common.Hash) *ParsedTx {
+	if idx, ok := p.TxHashes[hash]; ok {
+		return &p.Txs[idx]
+	}
+	return nil
+}
+
+// GetTxByMsgIndex returns ParsedTx by msg index
+func (p *ParsedTxs) GetTxByMsgIndex(i int) *ParsedTx {
+	if i < 0 || i >= len(p.Txs) {
+		return nil
+	}
+	return &p.Txs[i]
+}
+
+// GetTxByTxIndex returns ParsedTx by tx index
+func (p *ParsedTxs) GetTxByTxIndex(i int) *ParsedTx {
+	// assuming the `EthTxIndex` continuously increment.
+	if len(p.Txs) == 0 {
+		return nil
+	}
+	msgIndex := i - int(p.Txs[0].EthTxIndex)
```

Comment:
why is this always taking the same element? can you add a comment with the rationale?

Reply:
```
	// assuming the `EthTxIndex` increase continuously,
	// convert TxIndex to MsgIndex by substract the begin TxIndex.
```
added this comment.